### PR TITLE
feat: scib-autotune

### PR DESCRIPTION
**Pull request by ori-kron-wis**
re open the scib-autotune branch
see: https://github.com/scverse/scvi-tools/pull/2593
https://github.com/scverse/scvi-tools/pull/2517

### CODE REVIEW

File: src/scvi/train/_callbacks.py [hunk 1]
```diff
@@ -12,16 +12,26 @@
 import numpy as np
 import pyro
 import torch
+from anndata import AnnData
 from lightning.pytorch.callbacks import Callback, ModelCheckpoint
 from lightning.pytorch.callbacks.early_stopping import EarlyStopping
 from lightning.pytorch.utilities import rank_zero_info
+from scib_metrics.benchmark import (  # TODO: Need to be installed by default if we put it here
```
Comment (canergen) [inline src/scvi/train/_callbacks.py:19]: Needs to go into the CallBack. Don’t make it default.

File: src/scvi/train/_callbacks.py [hunk 2]
```diff
@@ -328,3 +338,60 @@ def on_train_start(self, trainer, pl_module):
...
+        if not hasattr(pl_module, f"_{stage}_epoch_outputs"):
+            raise ValueError(f"The training plan must have a `_{stage}_epoch_outputs` attribute.")
+
+        outputs = getattr(pl_module, f"_{stage}_epoch_outputs")
+        x = outputs["x"].numpy()
+        z = outputs["z"].numpy()
+        batch = outputs["batch"].numpy()
+        labels = outputs["labels"].numpy()
+
+        adata = AnnData(X=x, obs={"batch": batch, "labels": labels}, obsm={"z": z})
```
Comment (canergen) [inline src/scvi/train/_callbacks.py:376]: Doesn’t need X. We should be able to run it on some custom indices. It’s otherwise to expensive quickly.

Comment (ori-kron-wis) [reply src/scvi/train/_callbacks.py:376]: This is just an initial draft. No need to review yet.

File: src/scvi/autotune/_experiment.py [hunk 3]
```diff
@@ -26,6 +37,301 @@
...
+            ]
+        )
+
+
+    """
+
+    from scib_metrics.benchmark import BatchCorrection, BioConservation
+
+    def __init__(
+        self,
```
Comment (canergen) [inline src/scvi/autotune/_experiment.py:147]: Complete function documentation. What's num rows to select?

File: src/scvi/autotune/_experiment.py [hunk 4]
```diff
@@ -26,6 +37,301 @@
...
+            rand_idx = np.random.choice(z.shape[0], self.num_rows_to_select, replace=False)
+            batch = batch[rand_idx]
+            labels = labels[rand_idx]
+            x = x[rand_idx]
+            z = z[rand_idx]
+
+            # adjust which metric to run exactly
+            found_metric = next(
+                (key for key, value in metric_name_cleaner.items() if value == self.metric), None
+            )
```
Comment (canergen) [inline src/scvi/autotune/_experiment.py:219]: I don't get this line?

File: src/scvi/autotune/_experiment.py [hunk 5]
```diff
@@ -26,6 +37,301 @@
...
+            # special cases:
+            if self.metric == "Leiden NMI" or self.metric == "Leiden ARI":
```
Comment (canergen) [inline src/scvi/autotune/_experiment.py:221]: would be nice to change this directly in scib-metric. New argument for single metric.

File: src/scvi/autotune/_experiment.py [hunk 6]
```diff
@@ -90,6 +396,12 @@ class AutotuneExperiment:
         Additional keyword arguments to pass to the scheduler.
     searcher_kwargs
         Additional keyword arguments to pass to the search algorithm.
+    scib_stage
+        Used when performing scib-metrics tune, select whether to perform on validation (default)
+        or training end.
+    scib_subsample_rows
+        Used when performing scib-metrics tune, select number of rows to subsample (100 default).
```
Comment (canergen) [inline src/scvi/autotune/_experiment.py:403]: 100 should be to few. Should be 5k? Have you benchmarked it on real data

File: src/scvi/autotune/_experiment.py [hunk 7]
```diff
@@ -448,8 +764,30 @@ def metrics_callback(self) -> Callback:
...
+        # this is to get around lightning import changes
+        callback_cls = type(
+            "_ScibTuneReportCheckpointCallback",
+            (ScibTuneReportCheckpointCallback, Callback),
+            {},
+        )
+
+        return callback_cls(
+            metrics=self.metrics,
+            metric=self.metrics[0],
```
Comment (canergen) [inline src/scvi/autotune/_experiment.py:784]: What's the difference between metrics and metric?

File: src/scvi/autotune/_experiment.py [hunk 8]
```diff
@@ -524,20 +862,35 @@ def _trainable(
...
+    else:
+        tune_callback = [experiment.metrics_callback]
     train_params = {
         "accelerator": "auto",
         "devices": "auto",
         "check_val_every_n_epoch": 1,
         "enable_progress_bar": False,
         "logger": experiment.get_logger(get_context().get_trial_name()),
-        "callbacks": [experiment.metrics_callback],
+        "callbacks": tune_callback,  # change it to scib metrics callback here?
```
Comment (canergen) [inline src/scvi/autotune/_experiment.py:893]: remove comment. fine like this.

File: src/scvi/train/_trainingplans.py [hunk 9]
```diff
@@ -344,16 +344,82 @@ def compute_and_log_metrics(
...
+                sum(
+                    [
+                        "Scib" in x
+                        for x in [cls.__class__.__name__ for cls in self.trainer.callbacks]
+                    ]
+                )
+                > 0
+            ):
+                # TODO: subsample to save time already here?
+                x = loss_outputs["x"].detach().cpu()
```
Comment (canergen) [inline src/scvi/train/_trainingplans.py:384]: this can be very memory intensive.

Comment (canergen) [reply src/scvi/train/_trainingplans.py:384]: subsampling should be to the same cells in a ray tune run. Maybe expose the indices instead? It would also make it more flexible to e.g. set one single dataset as scib cells.

File: src/scvi/autotune/_experiment.py (hunk 6 above)
Comment (ori-kron-wis) [reply src/scvi/autotune/_experiment.py:403]: no. but of course 100 is low, Ill change that.

File: src/scvi/autotune/_experiment.py (hunk 3 above)
Comment (ori-kron-wis) [reply src/scvi/autotune/_experiment.py:147]: its the number of random rows we want to perform the scib metric calculation based on

File: src/scvi/autotune/_experiment.py (hunk 4 above)
Comment (ori-kron-wis) [reply src/scvi/autotune/_experiment.py:219]: actually we dont have to even do that and just stick to the metric name

File: src/scvi/train/_trainingplans.py (hunk 9 above)
Comment (ori-kron-wis) [reply src/scvi/train/_trainingplans.py:384]: We dont really need X. Ill change that and Ill add the option to input specific indices

File: src/scvi/autotune/_experiment.py (hunk 5 above)
Comment (ori-kron-wis) [reply src/scvi/autotune/_experiment.py:221]: waiting with this for now

File: src/scvi/train/_trainingplans.py (hunk 9 above)
Comment (ori-kron-wis) [reply src/scvi/train/_trainingplans.py]: also remember that this while mechanism works only if we set to scib tune

**Review by canergen — APPROVED** [review]
We should allow for custom indices to evaluate scib on (expose index argument and only  use random indices if it is None).
